x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/AnomalyDetectorsIndex.java

```diff
@@ -120,60 +138,99 @@ public static String configIndexName() {
     public static void createStateIndexAndAliasIfNecessary(Client client, ClusterState state, IndexNameExpressionResolver resolver,
                                                            final ActionListener<Boolean> finalListener) {
 
-        if (state.getMetaData().getAliasAndIndexLookup().containsKey(jobStateIndexWriteAlias())) {
-            finalListener.onResponse(false);
-            return;
-        }
+        String[] stateIndices = resolver.concreteIndexNames(state, IndicesOptions.lenientExpandOpen(), jobStateIndexPattern());
+        Optional<IndexMetaData> indexPointedByCurrentWriteAlias = state.getMetaData().hasAlias(".ml-state-write")
```

Comment:
What happens if `.ml-state-write` is a concrete index? I may have missed a logic branch somewhere.

Reply:
Our code never creates such a situation (i.e. a concrete index named `.ml-state-write`). It may occur if there is a bug in our code or the user creates `.ml-state-write` index before creating an ML job.
So I'm wondering what is the right thing to do in such a case. Is it even possible to have both index and alias named `.ml-state-write`? Let me find out...

Reply:
I verified that when an index ".ml-state-write" exists, one cannot create an alias with the same name:
```
{
  "error": {
    "root_cause": [
      {
        "type": "invalid_alias_name_exception",
        "reason": "Invalid alias name [.ml-state-write], an index exists with the same name as the alias",
        "index_uuid": "YyvpSSoSQc6yoBq69Q0gDg",
        "index": ".ml-state-write"
      }
    ],
    "type": "invalid_alias_name_exception",
    "reason": "Invalid alias name [.ml-state-write], an index exists with the same name as the alias",
    "index_uuid": "YyvpSSoSQc6yoBq69Q0gDg",
    "index": ".ml-state-write"
  },
  "status": 400
}
```

So I'd propose to leave the code as it is because this situation (pre-existing `.ml-state-write` index) is unlikely and could have as well happened before this PR.

Reply:
@droberts195: Did you give thumb up for leaving the code as it is or for throwing an exception?

Reply:
Sorry for the ambiguity.  I was giving thumbs up to

> So I'd propose to leave the code as it is

`.ml-state-write` should not really exist as a concrete index.  We have never used this name for a concrete index and if a user has then that's at the level of hacking the system in an unsupported way.  I am sure there are masses of other checks we could add for users creating indices that would confuse our code, but we haven't added such checks because we don't expect this sort of hacking and in practice people who ask for help haven't done this.

Reply:
Cool, thanks for confirming.